fix(mailer): Fix the bulk-mailer, add lots of tests.

shane-tomlinson opened the pull request:
fixes #2416

shane-tomlinson commented on scripts/bulk-mailer/bulk-mailer.js:
```diff
@@ -0,0 +1,33 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
```
Sure thing. Aren't you on PTO?

philbooth replied:
Yep, but I'm online, just happened to spot this when it arrived in my mailbox. I'm working on my vi-clone again!

vbudhram commented on scripts/bulk-mailer.js:
```diff
@@ -6,291 +6,48 @@
-  .option('-t, --template <template>', 'Template filename to render')
-  .option('-u, --unsent [filename]', 'JSON output file that contains emails that were not sent. Defaults to ./unsent.json')
+  .option('-i, --input <filename>', 'JSON user input file')
+  .option('-m, --method <method>', 'Sender method to call')
```
Something to consider, certain sender methods can send multiple template types. I don't think there is anything actionable but just wanted to call that out.

vbudhram commented on scripts/dump-users.js:
```diff
@@ -0,0 +1,81 @@
+
+
+function parseEmails(emails = '') {
+  if (! emails.length) {
```
Should we impose some limit on how many emails/uid can be dumped at once?

shane-tomlinson replied:
It would be a somewhat flimsy roadblock, a malicious user with access to the server would just run the script again or dump the DB by running a query.

vbudhram replied:
I was thinking it would be more for us, if we are trying to dump 1000 users and slow down database. Although, I guess if it is done on a read replica then it wouldn't really matter 🤷🏽‍♂️

shane-tomlinson commented on scripts/bulk-mailer/index.js:
```diff
@@ -0,0 +1,85 @@
+ * @param {Boolean} [shouldSend=false]
+ * @param {String} [emailOutputFilename=''] if specified, write emails to disk at this location
+ */
+module.exports = async function (userRecordFilename, mailerMethodName, batchSize, batchDelayMS, shouldSend, emailOutputDirname) {
```
Those defaults are set in scripts/bulk-mailer.js and not here, I'll remove the default values from the comments here since they no longer make sense.

shane-tomlinson commented on test/scripts/dump-users.js:
```diff
@@ -0,0 +1,190 @@
+    return DB.connect(config[config.db.backend])
+      .then(_db => {
+        db = _db;
+        return db.createAccount({
```
These tests fail on travis because it cannot connect to the DB. Work aroundable?

vbudhram commented on scripts/bulk-mailer/normalize-user-records.js:
```diff
@@ -0,0 +1,93 @@
+      .filter(record => !! record.email)
+      .map(userRecord => this.normalizeUserRecord(userRecord, translator))
+  }
+
```
nit: Extra space

vbudhram commented on scripts/dump-users.js:
```diff
@@ -0,0 +1,67 @@
+program
+  .option('-e, --emails [emails]', 'Email addresses to dump, comma separated')
+  .option('-u, --uids [uids]', 'User IDs to dump, comma separated')
+  .option('-i, --input <filename>', 'Input filename from which to read input if not specified on the command line')
```
👍

vbudhram commented on scripts/dump-users.js:
```diff
@@ -0,0 +1,67 @@
+
+
+const keys = emails.length ? emails : uids
+const dbFunc = emails.length ? 'emailRecord': 'account'
```
I think we should use `accountRecord` here instead of `emailRecord`. I remember trying to depreciate [emailRecord](https://github.com/mozilla/fxa-auth-server/issues/1999#issuecomment-316447124) but ran into some snags and punted on the idea. However, at the least I can put up a PR to add some notes that `emailRecord` is only to be used internally.

vbudhram commented on scripts/dump-users/index.js:
```diff
@@ -0,0 +1,89 @@
+      userRecord,
+      'createdAt',
+      'devices',
+      'email',
```
It is safe to filter out `email`,`emailVerified`, and `normalizedEmail` here too since they are stored in the `emails` object.

shane-tomlinson replied:
Do the templates ever make use of those?

vbudhram replied:
They do, but access it through the `emails` and `primaryEmail` object.

vbudhram commented on test/scripts/bulk-mailer.js:
```diff
@@ -0,0 +1,124 @@
+    return DB.connect(config[config.db.backend])
+      .then(_db => {
+        db = _db;
+        return db.createAccount({
```
Might make testing a little cleaner if you mock account object

```javascript
const accountMock = {
          email: 'user2@test.com',
          uid: 'f9916686c226415abd06ae550f073ced',
}
```

so you can reference `accountMock.uid` instead of the uid hex value.

vbudhram commented on test/scripts/bulk-mailer.js:
```diff
@@ -0,0 +1,124 @@
+
+  it('fails if --input missing', () => {
+    return cp.execAsync('node scripts/bulk-mailer --method sendVerifyCode', { cwd })
+      .then(() => assert(false, 'script should have failed'), () => {})
```
I think we might be able to check the output and assert correct error message for some of these.

jrgm commented on scripts/bulk-mailer/index.js:
```diff
@@ -0,0 +1,111 @@
+
+/**
+ * Send an email to users listed in the file `userRecordFilename` using `mailerMethodName`
+ * in batches of `batchSize` with `batchDelayMS` between batches.
```
I'm sending 28 messages in stage, with default batchsize (10) and delay (5000ms) and all 28 are sent in 2 to 6 seconds, so it doesn't look like this delay is working properly.

shane-tomlinson replied:
Turns out [I'm not making use of this value](https://github.com/mozilla/fxa-auth-server/blob/5c51b92e505b3842515840c47dee9110c3d26bcb/scripts/bulk-mailer/send-email-batches.js#L26). I converted from homegrown processing of the batches to use P.each on each of the batches and forgot to make use of the delay.

jrgm commented on scripts/bulk-mailer/index.js:
```diff
@@ -0,0 +1,111 @@
+  amplitudeEvent: () => {},
+  error: console.error,
+  info: () => {},
+  trace: () => {}
```
Log all the things!

Right now, when I run `./scripts/bulk-mailer.js` the only message is a `console.log('done')`.

This script should log each intent (`sending foo@example`), and result (`sent to foo@example.com at ${time} with SMTP Message-Id ${messageId}).

If there is some configuration setting I need to make, please let me know.

jrgm replied:
(Actually, I don't know what I'd do with the message-id, but it makes me warm and fuzzy that the providers response actually had assigned one).

jrgm commented on scripts/bulk-mailer/normalize-user-records.js:
```diff
@@ -0,0 +1,92 @@
+  }
+
+  normalizeAcceptLanguage(userRecord) {
+    // The Chinese translations were handed to us as "zh" w/o a country
```
For a user in stage that signed up with `accept-language: fr-FR`, when I send a bulk email to them, I'm not getting a translation in some types of email of interest. Is that expected?

shane-tomlinson replied:
This should now be fixed.

jrgm commented on scripts/bulk-mailer/send-email-batch.js:
```diff
@@ -0,0 +1,44 @@
+        .then(
+          () => {
+            successCount++
+            log.info({
```
`log.info` here must be a no-op here as I do not see this message (although if I stick a `console.log` right there, that is printed).

shane-tomlinson replied:
Fixed

jrgm commented:
I've installed this in stage on an fxa-auth instance. Using the environment file on that box as config, for 28 test accounts at a single email provider I've been able to generate the users.json, and use that input to send (and receive!) an email. 

I had some comments above, which summarized are:
- batch delay not working
- emails not being output to stdout or directory when --send is not used
- Log all the things! Log all the things! Log all the things!
- L10N templates

So looks good, modulo issues noted above.

Perhaps folks can give me the email of accounts set up in stage, and I can do a run that will send you emails at your email provider and you can review the content.

Also, one other bug: I used `--method sendVerifyCode` and the bulk-mailer crashed since it had an undefined verification code (and tried to call `toString` on it).

shane-tomlinson commented:
> batch delay not working

Fixed

> emails not being output to stdout or directory when --send is not used

Haven't been able to reproduce, can you give your STRs?

> Log all the things! Log all the things! Log all the things!

Fixed, add a `-v` on the command line.

> L10N templates

Still broken

shane-tomlinson commented:
> > L10N templates
> Still broken

Should now be fixed

jrgm commented on scripts/bulk-mailer/index.js:
```diff
@@ -0,0 +1,111 @@
+ * @param {Boolean} shouldSend if emails should be sent, for real. If false, emails
+ *   will instead be written to stdout or a directory. If `emailOutputDirname` is undefined,
+ *   emails will be written to stdout, if `emailOutputDirname` is defined, it will be used
+ *   as the output directory.
```
I don't see this at all running with `node ./scripts/bulk-mailer.js -i users.json -m revokeRecoveryKeyEmail`.

This is output in ./scripts/bulk-mailer/nodemailer-mocks/stream-output-mock.js and I see that constructor called, but not the method.

jrgm commented on scripts/bulk-mailer/index.js:
```diff
@@ -42,8 +36,25 @@ const logMock = {
+    error: console.error,
+    info (msg) {
+      if (useVerboseLogging) {
+        console.info(msg)
```
Can you wrap the `msg` to `JSON.stringify(msg)` so the messages are line oriented? Otherwise, it's an unparseable mix of single and multi-line messages.

jrgm commented:
> > batch delay not working
> 
> Fixed

Yes, appears to be working now. `--batchsize 2 --delay 10` takes 22 seconds, and `--batchsize 2 --delay 15` takes 32 seconds.
> 
> > emails not being output to stdout or directory when --send is not used
> 
> Haven't been able to reproduce, can you give your STRs?

Simply `node ./scripts/bulk-mailer.js -v -i users-5.json -m revokeRecoveryKeyEmail`.

It doesn't appear that `--send` matters. This always sends email, with or without `--send`. Yet the mail content output never happens. The mock method is never called.

> 
> > Log all the things! Log all the things! Log all the things!
> 
> Fixed, add a `-v` on the command line.

Thanks. I added a request to single-lineify the `msg` with JSON.stringify.

> 
> > L10N templates
> 
> Still broken

jrgm commented:
> Simply `node ./scripts/bulk-mailer.js -v -i users-5.json -m revokeRecoveryKeyEmail`

Sorry. `node ./scripts/bulk-mailer.js -v -i users-5.json --method sendVerifyCode`.

Also, real database.

vbudhram commented:
> It doesn't appear that --send matters. 

FWIW, tested this with `NODE_ENV=dev node ./scripts/bulk-mailer.js -v -i users.json --method sendVerifyCode --send` and without --send, and it appears to work as expected.

jrgm commented:
> It doesn't appear that --send matters. This always sends email, with or without --send. Yet the mail content output never happens. The mock method is never called.

This was due to having `email:config` in redis configured, and the select email services code path ignores the provided mock mailer.

philbooth commented on scripts/bulk-mailer/read-user-records.js:
```diff
@@ -0,0 +1,21 @@
+ * @param {String} userRecordsPath JSON file that contains user records
+ * @returns {Promise} resolves to an array of user records when complete
+ */
+module.exports = async function readUserRecords(userRecordsPath) {
```
...although I notice the tests are expecting a promise, so maybe it's deliberate. You can ignore me, sorry.
